File: backend/app/models/reflection.py
import uuid
from datetime import date
from datetime import datetime

from sqlalchemy import Date
from sqlalchemy import DateTime
from sqlalchemy import ForeignKey
from sqlalchemy import Text
from sqlalchemy import UniqueConstraint
from sqlalchemy import func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column

from app.db.base import Base


class Reflection(Base):
    """A written end-of-day reflection the user submits before work ends."""

    __tablename__ = "reflections"
    __table_args__ = (
        UniqueConstraint("user_id", "date", name="ux_reflections_user_date"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    date: Mapped[date] = mapped_column(
        Date,
        nullable=False,
        index=True,
    )

    text: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )

    analysis: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: ios/MyApp/Models/Reflection.swift
import Foundation

struct Reflection: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    let userId: UUID
    let date: Date
    let text: String
    let analysis: String?
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case date
        case text
        case analysis
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct ReflectionCreate: Encodable, Sendable {
    let date: Date
    let text: String
}

struct ReflectionAnalysisResponse: Codable, Hashable, Sendable {
    let id: UUID
    let userId: UUID
    let date: Date
    let analysis: String

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case date
        case analysis
    }
}
